### ISSUE TITLE
ProMP Heisenbug

### ISSUE THREAD

**Issue by AlexanderFabisch** [rock-learning/bolero#79]
We've got a [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug) in ProMPs. :)

See for example this build: https://circleci.com/gh/rock-learning/bolero/88, from build stage:

```
learning/bolero/src/representation/promp build error
...
Errors:
 - build: learning/bolero/src/representation/promp
```

However, unit tests of `ProMPBehavior` seem to be OK:

```
test_promp_behavior.test_smoke ... ok
test_promp_behavior.test_dmp_default_promp ... ok
test_promp_behavior.test_promp_get_set_params ... ok
test_promp_behavior.test_promp_constructor_args ... ok
test_promp_behavior.test_promp_imitate ... ok
```

I tried to reproduce it:
* pulled the docker image `af01/bolero_dependencies:latest`, cloned the repository, ran the [build code](https://github.com/rock-learning/bolero/blob/master/build_tools/circleci/build.sh); no error
* restarted the CircleCI build with SSH connection, removed the bolero-dev folder, ran bootstrap script again; no error

The error only seems to occur consistently during the first bootstrap in CircleCI. Even checked if that is an issue of pybob, but pybob only seems to report errors if the return code of make is `!= 0` which is not the case when I run it. Maybe @malter has another idea though.

**Comment by malter** [comment]
You can try to get log information by:
bob-show-log learning/bolero/src/representation/promp

**Comment by AlexanderFabisch** [comment]
True, this is the output:

```
bob-show-log learning/bolero/src/representation/promp
configure log:
-- Configuring done
-- Generating done
-- Build files have been written to: /root/project/bolero-dev/learning/bolero/src/representation/promp/build
build log:
make: Entering directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
make[1]: Entering directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
make[2]: Entering directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
make[2]: Leaving directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
[100%] Built target promp
make[2]: Entering directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
make[2]: Leaving directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
make[2]: Entering directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
running build_ext
building 'promp' extension
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c promp.cpp -o build/temp.linux-x86_64-2.7/promp.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/promp.cpp -o build/temp.linux-x86_64-2.7/../src/promp.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/BasisFunctions.cpp -o build/temp.linux-x86_64-2.7/../src/BasisFunctions.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/Trajectory.cpp -o build/temp.linux-x86_64-2.7/../src/Trajectory.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
c++ -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wl,-Bsymbolic-functions -Wl,-z,relro -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/promp.o build/temp.linux-x86_64-2.7/../src/promp.o build/temp.linux-x86_64-2.7/../src/BasisFunctions.o build/temp.linux-x86_64-2.7/../src/Trajectory.o -o /root/project/bolero-dev/learning/bolero/src/representation/promp/implementation/promp/promp.so
make[2]: Leaving directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
[100%] Built target python_setup
make[1]: Leaving directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
Install the project...
-- Install configuration: "DEBUG"
-- Installing: /root/project/bolero-dev/install/lib/libpromp.so
-- Installing: /root/project/bolero-dev/install/include/dmp/BasisFunctions.h
-- Installing: /root/project/bolero-dev/install/include/dmp/Trajectory.h
-- Installing: /root/project/bolero-dev/install/include/dmp/promp.h
running install
running build
running build_ext
building 'promp' extension
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c promp.cpp -o build/temp.linux-x86_64-2.7/promp.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/promp.cpp -o build/temp.linux-x86_64-2.7/../src/promp.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/BasisFunctions.cpp -o build/temp.linux-x86_64-2.7/../src/BasisFunctions.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/Trajectory.cpp -o build/temp.linux-x86_64-2.7/../src/Trajectory.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
creating build/lib.linux-x86_64-2.7
c++ -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wl,-Bsymbolic-functions -Wl,-z,relro -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/promp.o build/temp.linux-x86_64-2.7/../src/promp.o build/temp.linux-x86_64-2.7/../src/BasisFunctions.o build/temp.linux-x86_64-2.7/../src/Trajectory.o -o build/lib.linux-x86_64-2.7/promp.so
running install_lib
copying build/lib.linux-x86_64-2.7/promp.so -> /root/project/bolero-dev/install/lib/python2.7/site-packages
running install_egg_info
Removing /root/project/bolero-dev/install/lib/python2.7/site-packages/UNKNOWN-0.0.0-py2.7.egg-info
Writing /root/project/bolero-dev/install/lib/python2.7/site-packages/UNKNOWN-0.0.0-py2.7.egg-info
make: Leaving directory '/root/project/bolero-dev/learning/bolero/src/representation/promp/build'
Installed packages: 
[]
Time: 0:00:00.010276
```

@malter Following lines are printed in red. Here is my theory: pybob catches stderror and when it is not empty, it assumes that the installation fails? @bpoppinga Why is there anything on stderr? Why is it the case for ProMPs and not for DMPs?

```
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c promp.cpp -o build/temp.linux-x86_64-2.7/promp.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/promp.cpp -o build/temp.linux-x86_64-2.7/../src/promp.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/BasisFunctions.cpp -o build/temp.linux-x86_64-2.7/../src/BasisFunctions.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/Trajectory.cpp -o build/temp.linux-x86_64-2.7/../src/Trajectory.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
c++ -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wl,-Bsymbolic-functions -Wl,-z,relro -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/promp.o build/temp.linux-x86_64-2.7/../src/promp.o build/temp.linux-x86_64-2.7/../src/BasisFunctions.o build/temp.linux-x86_64-2.7/../src/Trajectory.o -o /root/project/bolero-dev/learning/bolero/src/representation/promp/implementation/promp/promp.so
...
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c promp.cpp -o build/temp.linux-x86_64-2.7/promp.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/promp.cpp -o build/temp.linux-x86_64-2.7/../src/promp.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/BasisFunctions.cpp -o build/temp.linux-x86_64-2.7/../src/BasisFunctions.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -UNDEBUG -I. -I/usr/lib/python2.7/dist-packages/numpy/core/include -I../src -I/usr/include/eigen3 -I/usr/include/python2.7 -c ../src/Trajectory.cpp -o build/temp.linux-x86_64-2.7/../src/Trajectory.o -std=c++11 -O3 -Wno-cpp -Wno-unused-function
...
c++ -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wl,-Bsymbolic-functions -Wl,-z,relro -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/promp.o build/temp.linux-x86_64-2.7/../src/promp.o build/temp.linux-x86_64-2.7/../src/BasisFunctions.o build/temp.linux-x86_64-2.7/../src/Trajectory.o -o build/lib.linux-x86_64-2.7/promp.so
```

**Comment by malter** [comment]
pybob checks the return value of `make install`. If it is not 0 it assumes the installation failed.

**Comment by AlexanderFabisch** [comment]
But there is no way to check the return value in the logfile... Could we log it somewhere?

**Comment by AlexanderFabisch** [comment]
Could be fixed with https://github.com/rock-learning/package_set/pull/3, we have to verify that!

**Comment by AlexanderFabisch** [comment]
> Could be fixed with [rock-learning/package_set#3](https://github.com/rock-learning/package_set/pull/3), we have to verify that!

oh, probably not because it is autoproj-related

**Comment by AlexanderFabisch** [comment]
This has been fixed. The CircleCI runner was not able to compile with more than one thread.
